### ISSUE TITLE
SQL Server 2005 - Assertion failure - causes BugCheck dump

### ISSUE THREAD

**Issue by snoopydo** [opserver/Opserver#134]
As of pulling down Opserver this morning.  When exiting monitoring either via closing VS2015 or killing IIS Express process.  SQL reports two alerts (16 & 20) and logs a BugCheck dump with the following query.  

``` log
*   Exception Address = 0000000002BB59C1 Module(sqlservr+0000000001BB59C1)                                       
*   Exception Code    = c0000005 EXCEPTION_ACCESS_VIOLATION                                                      
*   Access Violation occurred writing address 0000000000000008                                                   
* Input Buffer 510 bytes -                                                                                       
*              /* SQL\SQLInstance.Databases.cs@28 */   Select db.database_i                                      
*  d Id,         db.name Name,         db.state State,         db.compatibi                                      
*  lity_level CompatibilityLevel,         db.recovery_model RecoveryModel,                                       
*          db.page_verify_option PageVerifyOption,         db.log_reuse_wai                                      
*  t LogReuseWait,         db.user_access UserAccess,         db.is_fulltex                                      
*  t_enabled IsFullTextEnabled,         db.is_read_only IsReadOnly,                                              
*   db.is_read_committed_snapshot_on IsReadCommittedSnapshotOn,         db.                                      
*  snapshot_isolation_state SnapshotIsolationState,          (Cast(st.Total                                      
*  Size As Float)*8)/1024 TotalSizeMB,         (Cast(sr.RowSize As Float)*8                                      
*  )/1024 RowSizeMB,         (Cast(ss.StreamSize As Float)*8)/1024 StreamSi                                      
*  zeMB,         (Cast(sti.TextIndexSize As Float)*8)/1024 TextIndexSizeMB,                                      
*           Cast(logs.cntr_value as Float)/1024 LogSizeMB,         Cast(log                                      
*  u.cntr_value as Float)/1024 LogSizeUsedMB  From sys.databases db       L                                      
*  eft Join sys.dm_os_performance_counters logu On db.name = logu.instance_                                      
*  name And logu.counter_name LIKE N'Log File(s) Used Size (KB)%'        Le                                      
*  ft Join sys.dm_os_performance_counters logs On db.name = logs.instance_n                                      
*  ame And logs.counter_name LIKE N'Log File(s) Size (KB)%'        Left Joi                                      
*  n (Select database_id, Sum(Cast(size As Bigint)) TotalSize                                                    
*         From sys.master_files                 Group By database_id) st On                                      
*   db.database_id = st.database_id       Left Join (Select database_id, Su                                      
*  m(Cast(size As Bigint)) RowSize                     From sys.master_file                                      
*  s                    Where type = 0                 Group By database_id                                      
*  ) sr On db.database_id = sr.database_id       Left Join (Select database                                      
*  _id, Sum(Cast(size As Bigint)) StreamSize                     From sys.m                                      
*  aster_files                    Where type = 2                Group By da                                      
*  tabase_id) ss On db.database_id = ss.database_id       Left Join (Select                                      
*   database_id, Sum(Cast(size As Bigint)) TextIndexSize                                                         
*    From sys.master_files                    Where type = 4                                                     
*   Group By database_id) sti On db.database_id = sti.database_id ;   Selec                                      
*  t db.database_id DatabaseId,         db.name Name,          lb.type Last                                      
*  BackupType,         lb.backup_start_date LastBackupStartDate,         lb                                      
*  .backup_finish_date LastBackupFinishDate,         lb.backup_size LastBac                                      
*  kupSizeBytes,         lb.compressed_backup_size LastBackupCompressedSize                                      
*  Bytes,         lbmf.device_type LastBackupMediaDeviceType,         lbmf.                                      
*  logical_device_name LastBackupLogicalDeviceName,         lbmf.physical_d                                      
*  evice_name LastBackupPhysicalDeviceName,         fb.backup_start_date La                                      
*  stFullBackupStartDate,         fb.backup_finish_date LastFullBackupFinis                                      
*  hDate,         fb.backup_size LastFullBackupSizeBytes,         fb.compre                                      
*  ssed_backup_size LastFullBackupCompressedSizeBytes,         fbmf.device_                                      
*  type LastFullBackupMediaDeviceType,         fbmf.logical_device_name Las                                      
*  tFullBackupLogicalDeviceName,         fbmf.physical_device_name LastFull                                      
*  BackupPhysicalDeviceName    From sys.databases db         Left Outer Joi                                      
*  n (Select *                            From (Select backup_set_id,                                            
*                                      database_name,                                                            
*                     backup_start_date,                                                                         
*        backup_finish_date,                                         backup                                      
*  _size,                                         null compressed_backup_si                                      
*  ze,                                         media_set_id,                                                     
*                            type,                                                                               
*  Row_Number() Over(Partition By database_name Order By backup_start_date                                       
*  Desc) as rownum                                    From msdb.dbo.backups                                      
*  et) b                                                               Wher                                      
*  e rownum = 1) lb            On lb.database_name = db.name         Left O                                      
*  uter Join msdb.dbo.backupmediafamily lbmf           On lb.media_set_id =                                      
*   lbmf.media_set_id And lbmf.media_count = 1         Left Outer Join (Sel                                      
*  ect *                            Fro                                    
```

this appears to be from SQLInstance.Databases.cs line 200

**Comment by NickCraver** [comment]
Which version of .Net are you running on that machine?

**Comment by snoopydo** [comment]
The SQL Server is a 2008r2 sp1 machine, with .Net 4 installed.

**Comment by NickCraver** [comment]
Which version of .Net 4 are you on? Opserver needs 4.6+ to run properly given that we use a lot of the latest APIs.

**Comment by snoopydo** [comment]
I'm currently running Opserver in VS2015 community on my computer which has .Net 4.6 installed, the servers that I'm wanting to monitor have .Net 4.5

**Comment by NickCraver** [comment]
I'm trying to understand the scenario here - are you only running Opserver transiently, not in a persistent IIS process? It's not really designed to be brought up and down without error, since you're almost certain to be interrupting *something* among the myriad of polling/monitoring actions happening.

I am still curious though, where are you seeing these exceptions?

**Comment by snoopydo** [comment]
This is a low priority / background type project to replace powershell scripts & rrd at work for me at the moment, so I'm running it transiently under IIS Express when I get to work on it and stop/start it as I make config changes and experiment with options.  

This exception is logged on the sql server that I'm wanting to monitor, its logged by SQL into its own SQL Server Log (Server/Management/SQL Server Logs/.... in SQL Server Management Studio), and only after I stop the monitoring.

I find it strange to believe closing a connection to sql would cause it to do a dump like this, I can only guess there is something with the management queries/views used in this case and work load on the sql box?  The alert 16 kind of implies a timing issue.

**Comment by NickCraver** [comment]
@snoopydo sorry I dropped this - do you have an exact version # for SQL?

**Comment by snoopydo** [comment]
output from @@version

```
Microsoft SQL Server 2005 - 9.00.5057.00 (X64)   Mar 25 2011 13:33:31   Copyright (c) 1988-2005 Microsoft Corporation  Standard Edition (64-bit) on Windows NT 6.1 (Build 7601: Service Pack 1) 
```

**Comment by NickCraver** [comment]
So we narrow this down to the query and not the connection - can you try running the raw query and see if the bugcheck occurs? Here's the raw text version:

```SQL
Select db.database_id Id,
       db.name Name,
       db.state State,
       db.compatibility_level CompatibilityLevel,
       db.recovery_model RecoveryModel,
       db.page_verify_option PageVerifyOption,
       db.log_reuse_wait LogReuseWait,
       db.user_access UserAccess,
       db.is_fulltext_enabled IsFullTextEnabled,
       db.is_read_only IsReadOnly,
       db.is_read_committed_snapshot_on IsReadCommittedSnapshotOn,
       db.snapshot_isolation_state SnapshotIsolationState,
       (Cast(st.TotalSize As Float)*8)/1024 TotalSizeMB,
       (Cast(sr.RowSize As Float)*8)/1024 RowSizeMB,
       (Cast(ss.StreamSize As Float)*8)/1024 StreamSizeMB,
       (Cast(sti.TextIndexSize As Float)*8)/1024 TextIndexSizeMB,
       Cast(logs.cntr_value as Float)/1024 LogSizeMB,
       Cast(logu.cntr_value as Float)/1024 LogSizeUsedMB
From sys.databases db
     Left Join sys.dm_os_performance_counters logu On db.name = logu.instance_name And logu.counter_name LIKE N'Log File(s) Used Size (KB)%' 
     Left Join sys.dm_os_performance_counters logs On db.name = logs.instance_name And logs.counter_name LIKE N'Log File(s) Size (KB)%' 
     Left Join (Select database_id, Sum(Cast(size As Bigint)) TotalSize 
                  From sys.master_files 
              Group By database_id) st On db.database_id = st.database_id
     Left Join (Select database_id, Sum(Cast(size As Bigint)) RowSize 
                  From sys.master_files 
                 Where type = 0 
              Group By database_id) sr On db.database_id = sr.database_id
     Left Join (Select database_id, Sum(Cast(size As Bigint)) StreamSize 
                  From sys.master_files 
                 Where type = 2
              Group By database_id) ss On db.database_id = ss.database_id
     Left Join (Select database_id, Sum(Cast(size As Bigint)) TextIndexSize 
                  From sys.master_files 
                 Where type = 4
              Group By database_id) sti On db.database_id = sti.database_id;
```

**Comment by snoopydo** [comment]
Query runs without an issue via Management Studio.  This bug check appeared when I stopped opserver in visual studio and the sql server was restoring a shipped log for our largest db.  This server is a log shipping destination, other than the system dbs, all dbs are in a recovering state.

**Comment by NickCraver** [comment]
@snoopydo Are you able to repro with Opserver now, or only in that circumstance? A bugcheck at all definitely means something on the SQL Server side, but I haven't seen this exact issue before in 2005 and I don't see anything relating to it mentioned in the updates past yours (through `2005.90.5324.0`, mostly SP4 CU1, CU2 and CU3).

**Comment by snoopydo** [comment]
Looking good.  Just spent some time trying different combinations, trying to get it to fail again.  Nothing faulted, no warnings etc from SQL.
I'll leave one sql server in the mix to be monitored and keep an eye on it.

**Comment by NickCraver** [comment]
Closing out to cleanup.
